vortex-jni/src/errors.rs

```diff
@@ -1,34 +1,94 @@
 use jni::JNIEnv;
+use jni::objects::JObject;
+use jni::sys::{JNI_FALSE, jboolean, jobject};
 use vortex::error::VortexError;
 
-pub static ILLEGAL_ARGUMENT_CLASS: &str = "java/lang/IllegalArgumentException";
-pub static RUNTIME_EXC_CLASS: &str = "java/lang/RuntimeException";
-
-pub trait Throwable: Sized {
-    /// Throw the error with a particular class type.
-    fn throw(self, env: &mut JNIEnv, throw_class: &str);
+#[derive(Debug, thiserror::Error)]
+pub enum JNIError {
+    #[error("Vortex Error: {0}")]
+    Vortex(VortexError),
+    #[error("JNI Error: {0}")]
+    Custom(jni::errors::Error),
+}
 
-    fn throw_with_msg(self, env: &mut JNIEnv, throw_class: &str, msg: &str);
+impl From<jni::errors::Error> for JNIError {
+    fn from(error: jni::errors::Error) -> Self {
+        JNIError::Custom(error)
+    }
+}
 
-    fn throw_illegal_argument(self, env: &mut JNIEnv) {
-        self.throw(env, ILLEGAL_ARGUMENT_CLASS);
+impl From<VortexError> for JNIError {
+    fn from(error: VortexError) -> Self {
+        JNIError::Vortex(error)
     }
+}
+
+/// Types that have a reasonable default value to use
+/// across the FFI.
+pub trait JNIDefault {
+    fn jni_default() -> Self;
+}
+
+impl JNIDefault for () {
+    fn jni_default() -> Self {}
+}
 
-    fn throw_runtime(self, env: &mut JNIEnv, msg: &str) {
-        self.throw_with_msg(env, RUNTIME_EXC_CLASS, msg);
+impl JNIDefault for jboolean {
+    fn jni_default() -> Self {
+        JNI_FALSE
     }
 }
 
-impl Throwable for VortexError {
-    fn throw(self, env: &mut JNIEnv, throw_class: &str) {
-        let error_string = self.to_string();
-        env.throw_new(throw_class, error_string.as_str())
-            .expect("ThrowFailed");
+macro_rules! default_integer {
+    ($type:path) => {
+        impl JNIDefault for $type {
+            fn jni_default() -> Self {
+                -1
+            }
+        }
+    };
+}
+
+macro_rules! default_nan {
+    ($type:path) => {
+        impl JNIDefault for $type {
+            fn jni_default() -> Self {
+                <$type>::NAN
+            }
+        }
+    };
+}
+
+default_integer!(jni::sys::jbyte);
+default_integer!(jni::sys::jshort);
+default_integer!(jni::sys::jint);
+default_integer!(jni::sys::jlong);
+default_nan!(jni::sys::jfloat);
+default_nan!(jni::sys::jdouble);
+
+// All objectful types default to null.
+impl JNIDefault for jobject {
```

Comment:
`jstring, jbyteArray, ...` are all aliases for `jobject` so they all inherit this impl